README.md
# ed25519_dart

Pure Dart implementation of Ed25519 - public-key signature system.
For more information, please follow [here](https://ed25519.cr.yp.to).

In general, code mimics behaviour of original Python implementation
with some extensions from [ActiveState Code Recipes](https://github.com/ActiveState/code).
Code is not tested for security requirements, so it is good idea to use it
on trusted local machines!

## Usage

A simple usage example:
```dart   
    import 'package:ed25519_dart/ed25519_dart.dart';
    
    void main() {
      // Create new random secret key
      var sk = secretKey();
      print("Created random secret key with length - ${sk.length}");
      // Derive public key from secret key
      var pk = publicKey(sk);
      print("Derived public key with length ${pk.length} from secret key");
      // Create simple message from list
      var msg = bytesFromList([1, 2, 3, 4]);
      print("Created simple message - $msg");
      // Sign message with pk secret ley and public key
      var signature = sign(msg, sk, pk);
      print("Message signed by secret and public key");
      // Verify message signature with public key
      var isVerified = verifySignature(signature, msg, pk);
      print("Created signature is verified - $isVerified");
    }
```

## Package documentation
Extended package documentation can be found [here](https://www.dartdocs.org/documentation/ed25519_dart/0.0.1/).


CHANGELOG.md
# Changelog

## 1.0.2

- Release unused constants
- Type fixes

## 1.0.0

- Release version 1
- Analysis complete
- Described API

## 0.1.0

- Test coverage
- Explicit version bump to use Dart-like versions
- Added ability to generate secret key from seed
- Doc updates

## 0.0.1

- Initial version


test/ed25519_dart_test.dart
import 'dart:typed_data' show Uint8List;

import 'package:ed25519_dart/ed25519_dart.dart';
import 'package:test/test.dart';

void main() {
  test("bitClamp() clamps bits in provided lists as expected", () {
    var testData = new Uint8List(32);
    var testData2 = new Uint8List.fromList(
        [1, 2, 3, 4, 5, 6, 7, 8, 9, 10,
        11, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
        ]);
    var expected = new Uint8List.fromList(
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 64
        ]);
    var expected2 = new Uint8List.fromList(
        [0, 2, 3, 4, 5, 6, 7, 8, 9, 10,
        11, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 64
        ]
    );
    expect(bitClamp(testData), equals(expected));
    expect(bitClamp(testData2), equals(expected2));
  });

  test("bytesFromList() converts List into Uint8List", () {
    var testData = new List<int>.filled(32, 0);
    var expected = new Uint8List(32);
    expect(bytesFromList(testData), equals(expected));
    expect((bytesFromList(testData) is Uint8List), equals(true));
  });
  test("bytesToInteger() converts List of integers into integer", () {
    var testData = new List<int>.filled(32, 0);
    var testData2 = new List<int>.filled(32, 2);
    var expected = 0;
    var expected2 = 908173248920127022929968509872062022378588115024631874819275168689514742274;
    expect(bytesToInteger(testData), equals(expected));
    expect(bytesToInteger(testData2), equals(expected2));
  });
  test("decodePoint() creates expected point x-y point from passed integer", () {
    var testData = 1024128256;
    var expected = [5049901154188754798176685959377395864974690556190068451341045359400187598236, 1024128256];
    expect(decodePoint(testData), equals(expected));
  });
  test("edwards() adds two x-y points with expected result", () {
    var testData = [1,2];
    var expected = [38630462183868106874449484850498687242488459434265729964110945788313160992017,
                    2091706204615399755245742503625476174616215419841867942895091432328977195802];
    expect(edwards(testData, testData), equals(expected));
  });
  test("encodePoint() creates expected Uint8List from x-y point", () {
    var testData = [1,2];
    var expected = [2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 128];
    expect(encodePoint(testData), equals(expected));
  });
  test("Hash() hashes passed Uint8List into expected digest", () {
    var testData = new Uint8List(2);
    var expected = [94, 167, 29, 198, 208, 180, 245, 123, 243, 154, 173, 208,
                   124, 32, 140, 53, 240, 108, 210, 186, 197, 253, 226, 16, 57,
                   127, 112, 222, 17, 212, 57, 198, 46, 193, 205, 243, 24, 55,
                   88, 134, 95, 211, 135, 252, 234, 11, 173, 162, 246, 195, 122,
                   74, 23, 133, 29, 209, 215, 143, 239, 230, 242, 4, 238, 84];
    expect(Hash(testData), equals(expected));
  });
  test("integerToBytes() converts passed integer into expected Uint8List", () {
    var testData = 1024;
    var expected = [0, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];
    expect(integerToBytes(testData, 32), equals(expected));
    expect(integerToBytes(testData, 32).length, equals(32));
  });
  test("isOnCurve() checks if passed point is on curve", () {
    var testData = [15112221349535400772501151409588531511454012693041857206046113283949847762202,
                   46316835694926478169428394003475163141307993866256225615783033603165251855960];
    var testData2 = [1, 2];
    expect(isOnCurve(testData), equals(true));
    expect(isOnCurve(testData2), equals(false));
  });
  test("modularInverse() returns expected modular inverse", () {
    var testData = 2;
    var expected = 28948022309329048855892746252171976963317496166410141009864396001978282409975;
    expect(modularInverse(testData), equals(expected));
  });
  test("modularPow() computes and returns expected integer", () {
    var testData = 2;
    var expected = 256;
    expect(modularPow(testData, testData + 1), equals(expected));
  });

  test("publicKey() creates expected public key from passed secret key", () {
    var testData = new Uint8List(32);
    var expected = [59, 106, 39, 188, 206, 182, 164, 45, 98, 163, 168, 208, 42,
                   111, 13, 115, 101, 50, 21, 119, 29, 226, 67, 166, 58, 192,
                   72, 161, 139, 89, 218, 41];
    expect(publicKey(testData), equals(expected));
  });
  test("scalarMult() returns expected scalar multiplication of point", () {
    var testData = [1, 2];
    var expected = [2985072665741473095974513523900639186165599576139095168757154406154247914422,
                   48018218743393111964698292971632252982314570786198777847490644689840463716666];
    expect(scalarMult(testData, 10), equals(expected));
  });
  test("secretKey() returns random Uint8List with length 64", () {
    expect(secretKey().length, equals(64));
    expect((secretKey() is Uint8List), equals(true));
    expect(secretKey(1024).length, equals(64));
    expect((secretKey(1024) is Uint8List), equals(true));
  });
  test("sign() creates expected signature from passed parameters", () {
    var testData = new Uint8List(32);
    var expected = [62, 244, 112, 248, 0, 94, 55, 252, 113, 25, 122, 42, 208,
                   19, 144, 190, 229, 49, 127, 202, 143, 134, 163, 254, 26,
                   249, 204, 56, 245, 59, 169, 119, 177, 186, 224, 82, 94, 135,
                   255, 168, 240, 69, 169, 176, 231, 173, 200, 218, 4, 199, 186,
                   189, 225, 87, 155, 188, 192, 200, 82, 57, 113, 53, 213, 0];

    expect(sign(testData, testData, testData), equals(expected));
  });
  test("verifySignature() verifies signature by passed parameters", () {
    var testMessage = new Uint8List(32);
    var testPubKey = new Uint8List.fromList([57, 155, 96, 34, 234, 193, 54, 237, 157, 35, 24, 27, 183,
    158, 136, 254, 210, 211, 23, 93, 11, 51, 194, 141, 243, 85, 24, 152, 235,
    37, 55, 208]);
    var testSignature = new Uint8List.fromList([93, 80, 8, 206, 62, 55, 22, 162, 0, 162, 148, 32, 147,
                        9, 11, 227, 29, 64, 124, 140, 120, 58, 82, 207, 38, 112,
                        49, 145, 56, 131, 234, 108, 86, 147, 148, 31, 222, 23,
                        188, 196, 227, 234, 41, 234, 218, 8, 30, 93, 136, 117,
                        196, 32, 223, 180, 71, 106, 220, 89, 99, 116, 190, 214,
                        207, 0]);
    expect(verifySignature(testSignature, testMessage, testPubKey), equals(true));
  });
  test("xRecover() computes expected value from passed integer", () {
    var testData = 5;
    var expected = 39662079413846548812394406357884595898465868971068617279194812783497449546402;
    expect(xRecover(testData), equals(expected));
  });
}

lib/src/ed25519_dart_base.dart
/// Pure Dart implementation of Ed25519 - public-key signature system.
/// For more information, please follow https://ed25519.cr.yp.to.
/// In general, code mimics behaviour of original Python implementation
/// with some extensions from ActiveState Code Recipes
/// (Python 3 Ed25519 recipe - more at https://github.com/ActiveState/code).
/// Code is not tested for security requirements, so it is good idea to use it
/// on trusted local machines!

import 'dart:math' show Random, pow;
import 'dart:typed_data' show Uint8List;

import 'package:pointycastle/pointycastle.dart' show Digest;

// Magic constants
const int baseX =
15112221349535400772501151409588531511454012693041857206046113283949847762202;
const int baseY =
46316835694926478169428394003475163141307993866256225615783033603165251855960;
const int bits = 256;
const int d =
37095705934669439343138083508754565189542113879843219016388785533085940283555;
const int I =
19681161376707505956807079304988542015446066515923890162744021073123829784752;
const int primeL =
7237005577332262213973186563042994240857116359379907606001950938285454250989;
const int primeQ =
57896044618658097711785492504343953926634992332820282019728792003956564819949;

List<int> basePoint = const [baseX % primeQ, baseY % primeQ];

/// Clamps the lower and upper bits as required by the specification.
/// Returns [bytes] with clamped bits.
/// Length of the [bytes] should be at least 32.
///
///     var l = new List<int>.generate(32, (int i) => i + i); // [0, ..., 60, 62]
///     bitClamp(new Uint8List.fromList(l)); // [0, ..., 60, 126]
Uint8List bitClamp(Uint8List bytes) {
  bytes[0] &= 248;
  bytes[31] &= 63;
  bytes[31] |= 64;
  return bytes;
}

/// Returns [Uint8List] created from [lst].
/// Shortcut to avoid constructor duplication.
///
///     var bytes = bytesFromList([1, 2, 3]); // [1, 2, 3]
///     print(bytes.runtimeType); // Uint8List
Uint8List bytesFromList(List<int> lst) => new Uint8List.fromList(lst);

/// Converts [bytes] into fixed-size integer.
/// [bytes] length should be at least 32.
///
///     var l = new List<int>.generate(32, (int i) => i + i); // [0, ..., 60, 62]
///     bytesToInteger(l); // 28149809252802682310...81719888435032634998129152
int bytesToInteger(List<int> bytes) {
  num value = 0;
  bytes = bytes.sublist(0, 32);
  for (var i = 0; i < bytes.length; i++) {
    value += bytes[i] * pow(bits, i);
  }
  ;
  return value.toInt();
}

/// Converts integer [intVal] into [x, y] point.
///
///     decodePoint(28149809252802682310); // [2063...9514, 28149809252802682310]
List<int> decodePoint(int intVal) {
  var y = intVal % pow(2, (bits - 1)).toInt();
  var x = xRecover(y);
  if ((x & 1) != ((intVal >> (bits - 1))) & 1) {
    x = primeQ - x;
  }
  ;
  return [x, y];
}

/// Adds points on the Edwards curve.
/// Returns sum of points.
///
///     edwards([1,2], [1,2]); // [38630...2017, 20917...5802]
List<int> edwards(List<int> P, List<int> Q) {
  int x1, y1, x2, y2, x3, y3;
  x1 = P[0];
  y1 = P[1];
  x2 = Q[0];
  y2 = Q[1];
  x3 = (x1 * y2 + x2 * y1) * modularInverse(1 + d * x1 * x2 * y1 * y2);
  y3 = (y1 * y2 + x1 * x2) * modularInverse(1 - d * x1 * x2 * y1 * y2);
  return [x3 % primeQ, y3 % primeQ];
}

/// Encodes point [P] into [Uint8List].
///
///     encodePoint([1,2]); // [2, 0, ..., 0, 0, 128]
Uint8List encodePoint(List<int> P) {
  var x = P[0];
  var y = P[1];
  var encoded = integerToBytes(y + ((x & 1) << 255), 32);
  return encoded;
}

/// Returns digest message of SHA-512 hash function.
/// Digest message is result of hashing message [m].
///
///    Hash(new Uint8List(8)); // [27, 116, ..., 82, 196, 47, 27]
Uint8List Hash(Uint8List m) => new Digest('SHA-512').process(m);

/// Converts integer [e] into [Uint8List] with length [length].
///
///     integerToBytes(1, 32); // [0, 4, ... 0, 0, 0, 0, 0]
Uint8List integerToBytes(int e, int length) {
  var byteList = new Uint8List(length);
  for (var i = 0; i < length; i++) {
    byteList[0 + i] = (e >> (i * 8));
  }
  ;
  return byteList;
}

/// Returns [bool] that that indicates if point [P] is on curve.
///
///     isOnCurve([1, 2]); // false
bool isOnCurve(List<int> P) {
  int x, y;
  x = P[0];
  y = P[1];
  var onCurve = (-x * x + y * y - 1 - d * x * x * y * y) % primeQ == 0;
  return onCurve;
}

/// Returns the modular multiplicative inverse of integer [z]
/// and modulo [primeQ].
///
///     modularInverse(2); // 28948022...41009864396001978282409975
int modularInverse(int z) => z.modInverse(primeQ);

/// Returns integer [x] to the power of `pow(2, p)` with modulo [primeQ].
///
///     modularPow(3, 2); // 81
int modularPow(int x, int p) => x.modPow(pow(2, p).toInt(), primeQ);

/// Generates public key from given secret key [sk].
/// Public key is [Uint8List] with size 32.
///
///     publicKey(new Uint8List.fromList([1,2,3])); // [11, 198,162, ..., 184, 7]
Uint8List publicKey(Uint8List sk) {
  var skHash = Hash(sk);
  var clamped = bytesToInteger(bitClamp(skHash));
  var encoded = encodePoint(scalarMult(basePoint, clamped));
  return encoded;
}

/// Returns result of scalar multiplication of point [P] by integer [e].
///
///     scalarMult([1,2], 10); // [298...422, 480...666]
List<int> scalarMult(List<int> P, int e) {
  if (e == 0) {
    return [0, 1];
  }
  var Q = scalarMult(P, e ~/ 2);
  Q = edwards(Q, Q);
  if (e & 1 > 0) {
    Q = edwards(Q, P);
  }
  ;
  return Q;
}

/// Generates random secret key.
/// Accepts optional argument [seed] to generate random values.
/// When no arguments passed, then [Random.secure] is used to generate values.
/// Secret key is [Uint8List] with length 64.
///
///     secretKey(); // [224, 185, ..., 10, 17, 137]
///     secretKey(1024) // [225, 122, ..., 102, 232, 111]
Uint8List secretKey([int seed]) {
  Random randGen;
  if (seed == null) {
    randGen = new Random.secure();
  }
  else {
    randGen = new Random(seed);
  }
  var randList = new List<int>.generate(1024, (_) => randGen.nextInt(bits));
  var clamped = bitClamp(Hash(bytesFromList(randList)));
  return clamped;
}


/// Creates signature for message [message] by using secret key [secretKey]
/// and public key [pubKey].
/// Signature is [Uint8List] with size 64.
///
///     var m = new Uint8List(32);
///     var sk = new Uint8List(32);
///     var pk = new Uint8List(32);
///     sign(m, sk, pk); // [62, 244, 231, ..., 53, 213, 0]
Uint8List sign(Uint8List message, Uint8List secretKey, Uint8List pubKey) {
  var secretHash = Hash(secretKey);
  var secretKeyAddMsg = new List<int>.from(secretHash.sublist(32, 64));
  secretKeyAddMsg.addAll(message);
  var msgSecretAsInt = bytesToInteger(Hash(bytesFromList(secretKeyAddMsg)));
  var scalar = encodePoint(scalarMult(basePoint, msgSecretAsInt));
  var preDigest = new List<int>.from(scalar);
  preDigest.addAll(pubKey);
  preDigest.addAll(message);
  var digest = bytesToInteger(Hash(bytesFromList(preDigest)));
  var signature = new List<int>.from(scalar);
  signature.addAll(integerToBytes(
      (msgSecretAsInt + digest * bytesToInteger(bitClamp(secretHash))).toInt() %
          primeL,
      32));
  return bytesFromList(signature);
}

/// Verifies given signature [signature] with message [message] and
/// public key [pubKey].
/// Returns [bool] that indicates if verification is successful.
///
///     var sig = new Uint8List(32);
///     var m = new Uint8List(32);
///     var pk = new Uint8List(32);
///     verifySignature(sig, m, pk); // false
bool verifySignature(Uint8List signature, Uint8List message, Uint8List pubKey) {
  if (signature.lengthInBytes != bits / 4) {
    return false;
  }
  if (pubKey.length != bits / 8) {
    return false;
  }
  var sigSublist = signature.sublist(0, 32);
  var preDigest = new List<int>.from(sigSublist);
  preDigest.addAll(pubKey);
  preDigest.addAll(message);
  var hashInt = bytesToInteger(Hash(bytesFromList(preDigest)));
  var signatureInt = bytesToInteger(signature.sublist(32, 64));
  var signatureScalar = scalarMult(basePoint, signatureInt);
  var pubKeyScalar = scalarMult(decodePoint(bytesToInteger(pubKey)), hashInt);
  var edwardsPubKeyScalar =
  edwards(decodePoint(bytesToInteger(sigSublist)), pubKeyScalar);
  var verified = (signatureScalar.length == edwardsPubKeyScalar.length &&
      signatureScalar[0] == edwardsPubKeyScalar[0] &&
      signatureScalar[1] == edwardsPubKeyScalar[1]);
  return verified;
}

/// Recovers coordinate `x` by given coordinate [y].
/// Returns recovered `x`.
///
///     xRecover(10); // 246881771...00105170855113893569705867530
int xRecover(int y) {
  var xx = (y * y - 1) * modularInverse(d * y * y + 1);

  var x = xx.modPow((primeQ + 3) ~/ 8, primeQ);
  if ((x * x - xx) % primeQ != 0) {
    x = (x * I) % primeQ;
  }

  if (x % 2 != 0) {
    x = primeQ - x;
  }

  return x;
}
